doc/plugin_tutorial/tuto1/src/g_tuto1.mlg

```diff
@@ -8,149 +8,295 @@ DECLARE PLUGIN "tuto1_plugin"
-  functions that combine all the phases *)
-
-VERNAC COMMAND EXTEND TakingConstr2 CLASSIFIED AS QUERY
-| [ "Cmd2" constr(e) ] ->
-  { let _ = Constrintern.interp_constr
-    (Global.env())
-    (* Make sure you don't use Evd.empty here, as this does not
-      check consistency with existing universe constraints. *)
-    (Evd.from_env (Global.env())) e in
-    Feedback.msg_notice (strbrk "Cmd2 parsed something legitimate") }
+(*** Defining terms ***)
+
+(*
+ * To define a term, we start similarly to our intern functionality,
+ * then we call another function. We define this function in
+ * the Simple_declare module.
+ *
+ * The line #[ poly = Attributes.polymorphic ] says that this command accepts
+ * polymorphic attributes.
+ * @SkySkimmer: Here, poly is what the result is bound to in the
+ * rule's code. Multiple attributes may be used separated by ;, and we have
+ * punning so foo is equivalent to foo = foo.
+ *
+ * The declare_definition function returns the reference
+ * that was defined. This reference will be present in the new environment.
+ * If you want to refer to it later in your plugin, you must use an
+ * updated environment and the constructed reference.
+ *
+ * Note since we are now defining a term, we must classify this
+ * as a side-effect (CLASSIFIED AS SIDEFF).
+ *)
+VERNAC COMMAND EXTEND MyDefine CLASSIFIED AS SIDEFF
+| #[ poly = Attributes.polymorphic ] [ "MyDefine" ident(i) ":=" constr(e) ] ->
+   {
+     let env = Global.env () in
+     let sigma = Evd.from_env env in
+     let (sigma, t) = Constrintern.interp_constr_evars env sigma e in
+     let r = Simple_declare.declare_definition ~poly i sigma t in
+     let print r = strbrk "Defined " ++ Printer.pr_global r ++ strbrk "." in
+     Feedback.msg_notice (print r)
+   }
 END
 
-(* This is to show what happens when typing in an empty environment
-  with an empty evd.
- Question for the developers: why does "Cmd3 (fun x : nat => x)."
- raise an anomaly, not the same error as "Cmd3 (fun x : a => x)." *)
- 
-VERNAC COMMAND EXTEND TakingConstr3 CLASSIFIED AS QUERY
-| [ "Cmd3" constr(e) ] ->
-  { let _ = Constrintern.interp_constr Environ.empty_env
-      Evd.empty e in
-      Feedback.msg_notice
-        (strbrk "Cmd3 accepted something in the empty context")}
+(*** Printing terms ***)
+
+(*
+ * This command takes a name and return its value.  It does less
+ * than Print, because it fails on constructors, axioms, and inductive types.
+ * It signals an error to the user for unsupported terms.
+ *
+ * Simple_print contains simple_body_access, which shows how to look up
+ * a global reference.
+ *)
+VERNAC COMMAND EXTEND ExamplePrint CLASSIFIED AS QUERY
+| [ "MyPrint" reference(r) ] ->
+   {
+     let env = Global.env () in
+     let sigma = Evd.from_env env in
+     try
+       let t = Simple_print.simple_body_access (Nametab.global r) in
+       Feedback.msg_notice (Printer.pr_econstr_env env sigma t)
+     with Failure s ->
```

Comment:
Which function is raising this particular exception? It seems `Nametab.global` is raising `Not_found` instead.

Reply:
Yves' `simple_body_access` function in `simple_print.ml`.

Reply:
Then you also need to catch the `Not_found` otherwise you'll get anomalies (except if that's the desired behaviour).

Reply:
I tried with a non-existing reference, and instead of Not_found, I get:

```
Error: The reference foobar was not found in the current environment.
```

So I don't think this is a problem.

Reply:
Indeed, this is done at the genarg level.